src/shell_utils.cpp

```diff
@@ -0,0 +1,169 @@
+    }
+    return tokens;
+}
+
+// Executes a child process with the given command path and arguments
+void executeChildProcess(const std::string& commandPath, const std::vector<std::string>& arguments) {
+    std::vector<char*> args;
+    args.push_back(const_cast<char*>(commandPath.c_str())); // First argument is the command itself
+
+    // Prepare arguments for execvp
+    for (const auto& arg : arguments) {
+        args.push_back(const_cast<char*>(arg.c_str()));
+    }
+    args.push_back(nullptr); // Last argument must be nullptr
+
+    // Execute the command
+    if (execvp(commandPath.c_str(), args.data()) == -1) {
+        std::cerr << "Failed to execute command." << std::endl;
+        exit(EXIT_FAILURE);	// In case execvp fails
+    }
+}
+
+// Executes a command with the given command path, arguments, and background execution flag
+void executeCommand(const std::string& commandPath, const std::vector<std::string>& arguments, bool runInBackground) {
+    pid_t pid = fork();
+    if (pid < 0) {
+        std::cerr << "Fork failed." << std::endl;
+        return;
+    }
+    if (pid == 0) {
+        // Child process
+        executeChildProcess(commandPath, arguments);
+    }
+    else {
+        // Parent process
+        if (!runInBackground) {
+            int status;
+            waitpid(pid, &status, 0); // Wait for the child process to complete
+            // Update the job status based on the process exit status
+            bool isRunning = !WIFEXITED(status) || (WEXITSTATUS(status) == 0);
+            updateJobStatus(pid, isRunning);
+        }
+        else {
+            // Add the child process details (pid, command, etc.) to a list of background jobs
+            Job job(commandPath, pid, true);
+            backgroundJobs.push_back(job);
+        }
+    }
+}
+
+// Update the job status based on the process ID
+void updateJobStatus(int pid, bool isRunning) {
+    for (Job& job : backgroundJobs) {
+        if (job.pid == pid) {
+            job.isRunning = isRunning;
+            break;
+        }
+    }
+}
+
+// Prints the details of a job (command, PID, status)
+void printCommandDetails(const Job& job) {
+    std::cout << "Command: " << job.command << std::endl;
+    std::cout << "PID: " << job.pid << std::endl;
+    std::cout << "Status: " << (job.isRunning ? "Running" : "Finished") << std::endl;
+    std::cout << std::endl;
+}
+
+// Executes the 'myjobs' command to display information about background jobs
+void executeMyJobsCommand() {
+    if (backgroundJobs.empty()) {
+        std::cout << "No background jobs." << std::endl;
+        return;
+    }
+
+    // Iterate over the background jobs and print their details
+    for (const Job& job : backgroundJobs) {
+        printCommandDetails(job);
+    }
+}
+
+// Executes the 'cd' command to change the current working directory
+void executeCdCommand(const std::vector<std::string>& tokens) {
+    if (tokens.size() > 1) {
+        if (chdir(tokens[1].c_str()) != 0) {
```

Comment:
This way of changing dir is a bit old fashion, check this out (a more modern way): https://en.cppreference.com/w/cpp/filesystem/current_path